ssh/tailssh/incubator.go

```diff
@@ -670,14 +670,26 @@ func (ss *sshSession) launchProcess() error {
 
 	cmd := ss.cmd
 	homeDir := ss.conn.localUser.HomeDir
-	if _, err := os.Stat(homeDir); err == nil {
-		cmd.Dir = homeDir
-	} else if os.IsNotExist(err) {
-		// If the home directory doesn't exist, we can't chdir to it.
-		// Instead, we'll chdir to the root directory.
+	// We prefer to execute the child process in the connecting user's home
+	// directory. We need to be able to chdir to that directory in order to do
+	// so. How to check this? We don't want to chdir the parent tailscaled
+	// process, because that could have unexpected side-effects depending on
+	// what assumptions tailscaled makes about its working directory.
```

Comment:
We can't just try to chdir to it after we've dropped root?

This ReadDir(1) is still happening while we're root, right? If so, how does that help determine whether the dropped user will be able to access it?

Reply:
The incubator process is run as `root` and chdirs as `root`. We only drop privileges in those code branches where we don't launch `login` or `su`.

That said, maybe it makes sense only to chdir at the time that we drop privileges...

Reply:
Never mind, we `exec` both `login` and `su`, so we have to `chdir` before then, and we have to do it while still a privileged user.

Reply:
> how does that help determine whether the dropped user will be able to access it

What we're primarily interested in is whether whoever is forking the incubator process (typically root) is able to `chdir` to it. The OG problem is that we assumed that if we can stat the directory, we can chdir to it. In some cases, that's not true, so forking the incubator fails completely.

With this new approach, there is still the possibility that `root` can read the user's home directory but the user themselves can't, but I suspect that this is a misconfiguration that would cause all manner of problems outside of Tailscale SSH.